### ★ OVERSAMPLING

Oversampling is a way of making a dataset more balanced. This is done when there are too few examples of one type of item (minority class) in a dataset compared to another type (majority class). To balance the dataset, more examples of the minority class are added either by copying the existing ones or by making new ones. This helps make machine learning algorithms work better.

### ★ IMPORTED LIBRARIES

In [1]:
import tensorflow as tf
import pandas as pd
import numpy as np
import sklearn.linear_model
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from tensorflow import keras
from tensorflow.keras import layers
from tensorflow.keras.models import Sequential

2023-01-30 19:00:56.900376: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.


In [2]:
df = pd.read_csv("cleandf.csv")

In [3]:
df

Time         V1         V2        V3        V4        V5  \
0            0.0  -1.359807  -0.072781  2.536347  1.378155 -0.338321   
1            0.0   1.191857   0.266151  0.166480  0.448154  0.060018   
2            1.0  -1.358354  -1.340163  1.773209  0.379780 -0.503198   
3            1.0  -0.966272  -0.185226  1.792993 -0.863291 -0.010309   
4            2.0  -1.158233   0.877737  1.548718  0.403034 -0.407193   
...          ...        ...        ...       ...       ...       ...   
283721  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
283722  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
283723  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
283724  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
283725  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
0       0.462388  0.239599  0.098698  0.363787  ... -0.018307  0.277838   
1      -0.082361 -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672   
2       1.800499  0.791461  0.247676 -1.514654  ...  0.247998  0.771679   
3       1.247203  0.237609  0.377436 -1.387024  ... -0.108300  0.005274   
4       0.095921  0.592941 -0.270533  0.817739  ... -0.009431  0.798278   
...          ...       ...       ...       ...  ...       ...       ...   
283721 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
283722  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
283723  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
283724  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
283725 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
0      -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053  149.62   
1       0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69   
2       0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  378.66   
3      -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458  123.50   
4      -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153   69.99   
...          ...       ...       ...       ...       ...       ...     ...   
283721  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
283722  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
283723 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
283724 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
283725  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
0           0  
1           0  
2           0  
3           0  
4           0  
...       ...  
283721      0  
283722      0  
283723      0  
283724      0  
283725      0  

[283726 rows x 31 columns]

In [4]:
# Adding new column to categorize binary inputs into strings

df["Transaction"] = np.where(df["Class"]== 0, "Valid", "Fraud")

### ★ DATA INFORMATION

In [5]:
# Categorizing data 
valid = df[df.Class == 0]
fraud = df[df.Class == 1]

In [6]:
df.Class.value_counts()

0    283253
1       473
Name: Class, dtype: int64

In [7]:
# Defining "samplefraud" variable to produce same sample amount as valid data.

samplefraud = fraud.sample(len(valid), replace = True)

print("\n",'\033[1m' + 'Fraud transaction data shape' + '\033[0m')
print("\n",samplefraud.shape)


 Fraud transaction data shape

 (283253, 32)


In [8]:
# Making new dataset with 566,506 entries

df_new= pd.concat([valid , samplefraud], axis=0)

In [9]:
df_new['Class'].value_counts()

0    283253
1    283253
Name: Class, dtype: int64

In [10]:
df_new.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94835.058093  0.013439 -0.009829  0.012853 -0.010440  0.006769   
1      80419.426530 -4.502902  3.406938 -6.735873  4.470931 -2.957892   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.001251  0.010447 -0.002448  0.002613  ... -0.000489 -0.001150   
1     -1.436530 -5.187676  0.950499 -2.521099  ...  0.408072  0.463482   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000160  0.000360  0.000393 -0.000301  0.000065  0.001409  0.000418   
1      0.088448 -0.096779 -0.107438  0.038528  0.049947  0.215457  0.078929   

           Amount  
Class              
0       88.413575  
1      123.782537  

[2 rows x 30 columns]

### ★ LOGISTIC REGRESSION

In [11]:
x = df_new.drop(columns=['Class', "Transaction"], axis=1)
y = df_new['Class']

In [12]:
# Splitting data into test and train

X_training_data, X_testing_data, y_training_data, y_testing_data = train_test_split(x, y, test_size=0.2,stratify=y,random_state=42)


# Checking data distribution form original x into x train and x test

print('\033[1m' + 'X data shape' + '\033[0m',x.shape)
print('\033[1m' + 'X train data shape' + '\033[0m',X_training_data.shape)
print('\033[1m' + 'X test data shape' + '\033[0m',X_testing_data.shape)

X data shape (566506, 30)
X train data shape (453204, 30)
X test data shape (113302, 30)


In [13]:
logreg = LogisticRegression(max_iter = 4000,class_weight='balanced')
logreg.fit(X_training_data, y_training_data)
y_pred = logreg.predict(X_testing_data)

In [14]:
# Checking accuracy scores for train and test dataset

y_pred = logreg.predict(X_testing_data)
print('Accuracy of logistic regression classifier on train set: {:.2f}'.format(logreg.score(X_training_data, y_training_data)))
print('Accuracy of logistic regression classifier on test set: {:.2f}'.format(logreg.score(X_testing_data, y_testing_data)))

Accuracy of logistic regression classifier on train set: 0.94
Accuracy of logistic regression classifier on test set: 0.94


In [15]:
# Checking classification report results

print("\n", "1 = Fraud","  0 = Valid")
print("\n", '\033[1m' + 'CLASSIFICATION REPORT' + '\033[0m')
print(classification_report(y_testing_data,y_pred))


 1 = Fraud   0 = Valid

 CLASSIFICATION REPORT
              precision    recall  f1-score   support

           0       0.92      0.97      0.94     56651
           1       0.96      0.92      0.94     56651

    accuracy                           0.94    113302
   macro avg       0.94      0.94      0.94    113302
weighted avg       0.94      0.94      0.94    113302



In [16]:
# Checking confusion matrix results

confused = confusion_matrix(y_testing_data,y_pred)
print('\033[1m' + 'CONFUSION MATRIX:' + '\033[0m')
print("\n", confused)


CONFUSION MATRIX:

 [[54716  1935]
 [ 4498 52153]]


### ★ TENSORFLOW

TensorFlow is another approach that is good for datasets due to its wide range of pre-built models and algorithms, including deep learning models. Additionally, It also has efficient support for GPU processing, making it possible to train large models quickly.

In [17]:
# Building a very simple model with dense layer of 30 units is worked upon by a ReLU activation function

model = keras.Sequential([
    keras.layers.Dense(30, input_dim=30, activation='relu'),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dense(1, activation='sigmoid')])

2023-01-30 19:01:13.369314: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.


In [18]:
# Optimzer - 'adams'-- uses less memory and accelerates the gradient descent by looking at exponential weighted average.
# Loss - 'BinaryCrossentropy' -- compares each of the predicted probabilities to actual class output which can be either 0 or 1.
# Metrics - 'accuracy' -- Calculates the times when prediction equals the class names.

model.compile(optimizer='adam',
  loss=tf.keras.losses.BinaryCrossentropy(from_logits=True),
  metrics=['accuracy'])

In [19]:
# Defining variables

epochs = 250
batch_size = 4000

In [20]:
# Fitting the training dataset and validation dataset with their corresponding batch size and epochs

history =model.fit(X_training_data,y_training_data,validation_split = 0.20,batch_size = batch_size,epochs = epochs)

Epoch 1/250


/Users/sarbjitmadra/opt/anaconda3/lib/python3.9/site-packages/keras/backend.py:5676: UserWarning: "`binary_crossentropy` received `from_logits=True`, but the `output` argument was produced by a Sigmoid activation and thus does not represent logits. Was this intended?
  output, from_logits = _get_logits(


91/91 [==============================] - 1s 4ms/step - loss: 237.4112 - accuracy: 0.5117 - val_loss: 10.3705 - val_accuracy: 0.6167
Epoch 2/250
55/91 [=================>............] - ETA: 0s - loss: 36.7949 - accuracy: 0.5485

/Users/sarbjitmadra/opt/anaconda3/lib/python3.9/site-packages/keras/backend.py:5676: UserWarning: "`binary_crossentropy` received `from_logits=True`, but the `output` argument was produced by a Sigmoid activation and thus does not represent logits. Was this intended?
  output, from_logits = _get_logits(


91/91 [==============================] - 0s 2ms/step - loss: 34.2298 - accuracy: 0.5522 - val_loss: 29.5365 - val_accuracy: 0.5019
Epoch 3/250
91/91 [==============================] - 0s 3ms/step - loss: 29.3690 - accuracy: 0.5823 - val_loss: 22.3436 - val_accuracy: 0.6072
Epoch 4/250
91/91 [==============================] - 0s 2ms/step - loss: 29.4452 - accuracy: 0.5829 - val_loss: 18.2577 - val_accuracy: 0.6224
Epoch 5/250
91/91 [==============================] - 0s 3ms/step - loss: 21.5431 - accuracy: 0.5977 - val_loss: 29.2542 - val_accuracy: 0.5028
Epoch 6/250
91/91 [==============================] - 0s 3ms/step - loss: 19.6986 - accuracy: 0.6079 - val_loss: 28.0906 - val_accuracy: 0.5033
Epoch 7/250
91/91 [==============================] - 0s 3ms/step - loss: 18.4208 - accuracy: 0.6114 - val_loss: 13.0676 - val_accuracy: 0.7014
Epoch 8/250
91/91 [==============================] - 0s 2ms/step - loss: 17.2922 - accuracy: 0.6184 - val_loss: 27.4329 - val_accuracy: 0.6692
Epoch 9/250

91/91 [==============================] - 0s 2ms/step - loss: 16.9624 - accuracy: 0.7327 - val_loss: 17.7282 - val_accuracy: 0.7822
Epoch 60/250
91/91 [==============================] - 0s 2ms/step - loss: 4.5281 - accuracy: 0.8565 - val_loss: 5.3112 - val_accuracy: 0.8884
Epoch 61/250
91/91 [==============================] - 0s 2ms/step - loss: 1.5915 - accuracy: 0.9102 - val_loss: 1.0382 - val_accuracy: 0.9331
Epoch 62/250
91/91 [==============================] - 0s 2ms/step - loss: 1.0597 - accuracy: 0.9163 - val_loss: 0.9233 - val_accuracy: 0.9018
Epoch 63/250
91/91 [==============================] - 0s 3ms/step - loss: 9.0836 - accuracy: 0.8280 - val_loss: 8.3026 - val_accuracy: 0.5728
Epoch 64/250
91/91 [==============================] - 0s 3ms/step - loss: 15.7201 - accuracy: 0.7528 - val_loss: 1.4949 - val_accuracy: 0.9252
Epoch 65/250
91/91 [==============================] - 0s 3ms/step - loss: 0.9609 - accuracy: 0.9235 - val_loss: 0.7912 - val_accuracy: 0.9286
Epoch 66/250
91/

91/91 [==============================] - 0s 3ms/step - loss: 5.4933 - accuracy: 0.8637 - val_loss: 2.7531 - val_accuracy: 0.9259
Epoch 117/250
91/91 [==============================] - 0s 3ms/step - loss: 3.6380 - accuracy: 0.8836 - val_loss: 0.9089 - val_accuracy: 0.9341
Epoch 118/250
91/91 [==============================] - 0s 3ms/step - loss: 4.0620 - accuracy: 0.8818 - val_loss: 4.2062 - val_accuracy: 0.9102
Epoch 119/250
91/91 [==============================] - 0s 3ms/step - loss: 4.0623 - accuracy: 0.8799 - val_loss: 4.3221 - val_accuracy: 0.7339
Epoch 120/250
91/91 [==============================] - 0s 3ms/step - loss: 2.8479 - accuracy: 0.8967 - val_loss: 0.9388 - val_accuracy: 0.9289
Epoch 121/250
91/91 [==============================] - 0s 3ms/step - loss: 1.9171 - accuracy: 0.9146 - val_loss: 2.9297 - val_accuracy: 0.7936
Epoch 122/250
91/91 [==============================] - 0s 2ms/step - loss: 7.3304 - accuracy: 0.8456 - val_loss: 1.4033 - val_accuracy: 0.8967
Epoch 123/250

91/91 [==============================] - 0s 3ms/step - loss: 2.7270 - accuracy: 0.8739 - val_loss: 6.8823 - val_accuracy: 0.8686
Epoch 174/250
91/91 [==============================] - 0s 2ms/step - loss: 0.9315 - accuracy: 0.9252 - val_loss: 0.5805 - val_accuracy: 0.9413
Epoch 175/250
91/91 [==============================] - 0s 3ms/step - loss: 0.5421 - accuracy: 0.9331 - val_loss: 0.6381 - val_accuracy: 0.9430
Epoch 176/250
91/91 [==============================] - 0s 2ms/step - loss: 4.6797 - accuracy: 0.8489 - val_loss: 1.3394 - val_accuracy: 0.9287
Epoch 177/250
91/91 [==============================] - 0s 2ms/step - loss: 2.4777 - accuracy: 0.8904 - val_loss: 0.5596 - val_accuracy: 0.9397
Epoch 178/250
91/91 [==============================] - 0s 3ms/step - loss: 0.5234 - accuracy: 0.9354 - val_loss: 0.5126 - val_accuracy: 0.9324
Epoch 179/250
91/91 [==============================] - 0s 3ms/step - loss: 0.6932 - accuracy: 0.9283 - val_loss: 1.3442 - val_accuracy: 0.9290
Epoch 180/250

91/91 [==============================] - 0s 3ms/step - loss: 1.2029 - accuracy: 0.9230 - val_loss: 0.9971 - val_accuracy: 0.9401
Epoch 231/250
91/91 [==============================] - 0s 3ms/step - loss: 4.1806 - accuracy: 0.8687 - val_loss: 1.3334 - val_accuracy: 0.9369
Epoch 232/250
91/91 [==============================] - 0s 3ms/step - loss: 1.9524 - accuracy: 0.9050 - val_loss: 3.5566 - val_accuracy: 0.9132
Epoch 233/250
91/91 [==============================] - 0s 3ms/step - loss: 0.9233 - accuracy: 0.9288 - val_loss: 0.5792 - val_accuracy: 0.9373
Epoch 234/250
91/91 [==============================] - 0s 3ms/step - loss: 0.5828 - accuracy: 0.9372 - val_loss: 0.5879 - val_accuracy: 0.9315
Epoch 235/250
91/91 [==============================] - 0s 3ms/step - loss: 1.5826 - accuracy: 0.9206 - val_loss: 10.9402 - val_accuracy: 0.5549
Epoch 236/250
91/91 [==============================] - 0s 3ms/step - loss: 4.4189 - accuracy: 0.8668 - val_loss: 12.4574 - val_accuracy: 0.8417
Epoch 237/2

In [21]:
# Evaluating model's performance on the test dataset

print(model.evaluate(X_testing_data, y_testing_data))
    
y_preds = model.predict(X_testing_data)
y_preds = np.round(y_preds)


3541/3541 [==============================] - 3s 875us/step - loss: 0.6245 - accuracy: 0.9322
[0.6245129108428955, 0.9321900606155396]
3541/3541 [==============================] - 3s 723us/step


In [22]:
# Checking classification report results

print("\n", "1 = Fraud","  0 = Valid")
print("\n", '\033[1m' + 'CLASSIFICATION REPORT' + '\033[0m')
    
print("\n", classification_report(y_testing_data, y_preds))


 1 = Fraud   0 = Valid

 CLASSIFICATION REPORT

               precision    recall  f1-score   support

           0       0.95      0.91      0.93     56651
           1       0.91      0.96      0.93     56651

    accuracy                           0.93    113302
   macro avg       0.93      0.93      0.93    113302
weighted avg       0.93      0.93      0.93    113302



In [23]:
# Checking confusion matrix results

confused = confusion_matrix(y_testing_data,y_pred)
print('\033[1m' + 'CONFUSION MATRIX:' + '\033[0m')
print("\n", confused)

CONFUSION MATRIX:

 [[54716  1935]
 [ 4498 52153]]


### ★ CONCLUSION

Logistic regression and a TensorFlow Keras model were used for credit card fraud detection with an accuracy of 94% and 93% respectively. The data was oversampled to address class imbalance, resulting in 98% more data than its original size. However, oversampling can lead to overfitting, making the model less effective in classifying new and unseen data where the model becomes too specialized to the training data, and fails to generalize well to new data.

In this case, the high accuracy obtained through oversampling might not reflect the actual performance of the model in real-world scenarios, and caution should be taken when using this approach.